File: Samples VB.NET/Advanced/CategoryDesigner/CaseDefFieldViewModel.vb
Imports System.Globalization
Imports System.Text
Imports Therefore.API

Namespace Therefore.Samples
	Public Class CaseDefFieldViewModel
		Inherits FieldViewModel
		Private _casedef As TheCaseDefinition

		Public Sub New(ByVal cat As TheCaseDefinition)
			MyBase.New()
			_casedef = cat
			UpdateFieldList()
			UpdateLanguages()
		End Sub

		Public Overrides Sub UpdateFieldList()
			Dim fields As New List(Of TheCategoryField)()
			For i As Integer = 0 To _casedef.FieldCount - 1
				fields.Add(_casedef.GetFieldByIndex(i))
			Next i
			_filteredFields = CollectionViewSource.GetDefaultView(fields)
			_filteredFields.Filter = New Predicate(Of Object)(AddressOf FieldFilter)
			OnPropertyChanged("VisibleFields")
		End Sub

		Public Overrides Property CurrentLanguage() As CultureInfo
			Get
				If _languages.ContainsKey(_casedef.GetCurrentLCID()) Then
					Return _languages(_casedef.GetCurrentLCID())
				End If
				Return _languages.Values(0)
			End Get

			Set(ByVal value As CultureInfo)
				_casedef.SetCurrentLCID(value.LCID)
			End Set
		End Property

		Public Sub UpdateLanguages()
			_languages = New SortedList(Of Integer, CultureInfo)()
			Dim lcids As TheIntArray = _casedef.GetAvailableLCIDs()
			For Each lcid As Integer In lcids
				_languages.Add(lcid, New CultureInfo(lcid))
			Next lcid
		End Sub

		Public ReadOnly Property CaseDefinition() As TheCaseDefinition
			Get
				Return _casedef
			End Get
		End Property
	End Class
End Namespace

File: Samples CSharp/Advanced/CategoryDesigner/CaseDefFieldViewModel.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text;
using System.Windows.Data;
using Therefore.API;

namespace Therefore.Samples
{
    public class CaseDefFieldViewModel : FieldViewModel
    {
        private TheCaseDefinition _casedef;

        public CaseDefFieldViewModel(TheCaseDefinition cat) : base()
        {
            _casedef = cat;
            UpdateFieldList();
            UpdateLanguages();
        }

        override public void UpdateFieldList()
        {
            List<TheCategoryField> fields = new List<TheCategoryField>();
            for (int i = 0; i < _casedef.FieldCount; i++ )
                fields.Add(_casedef.GetFieldByIndex(i));
            _filteredFields = CollectionViewSource.GetDefaultView(fields);
            _filteredFields.Filter = new Predicate<object>(FieldFilter);
            OnPropertyChanged("VisibleFields");
        }

        override public CultureInfo CurrentLanguage
        {
            get
            {
                if (_languages.ContainsKey(_casedef.GetCurrentLCID())) 
                    return _languages[_casedef.GetCurrentLCID()];
                return _languages.Values[0];
            }

            set
            {
                _casedef.SetCurrentLCID(value.LCID);
            }
        }

        public void UpdateLanguages()
        {
            _languages = new SortedList<int, CultureInfo>();
            TheIntArray lcids = _casedef.GetAvailableLCIDs();
            foreach (int lcid in lcids)
                _languages.Add(lcid, new CultureInfo(lcid));
        }

        public TheCaseDefinition CaseDefinition
        {
            get { return _casedef; }
        }
    }
}
